File: python/circular-buffer/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """
    def __init__(self, message):
        pass


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """
    def __init__(self, message):
        pass


class CircularBuffer:
    def __init__(self, capacity):
        self.buffer = [None] * capacity
        self.read_index = 0
        self.write_index = 0

    def read(self):
        data = self.buffer[self.read_index]

        if not data:
            raise BufferEmptyException("Circular buffer is empty")

        self.buffer[self.read_index] = None
        self.read_index = (self.read_index + 1) % len(self.buffer)

        return data

    def write(self, data):
        if self.buffer[self.write_index]:
            raise BufferFullException("Circular buffer is full")

        self.buffer[self.write_index] = data
        self.write_index = (self.write_index + 1) % len(self.buffer)

    def overwrite(self, data):
        if self.buffer[self.write_index]:
            self.buffer[self.read_index] = data
            self.read_index = (self.read_index + 1) % len(self.buffer)
            self.write_index = self.read_index
        else:
            self.write(data)

    def clear(self):
        self.buffer = [None] * len(self.buffer)
        self.read_index = 0
        self.write_index = 0

File: javascript/circular-buffer/circular-buffer.js
class CircularBuffer {
  constructor(size) {
    this.buffer = new Array(size).fill(null);
    this.size = size;
    this.start = 0;
    this.end = 0;
    this.full = false;
  }

  write(value) {
    if (this.full) {
      throw new BufferFullError();
    }

    this.buffer[this.end] = value;
    this.end = (this.end + 1) % this.size;

    if (this.end === this.start) {
      this.full = true;
    }
  }

  read() {
    if (this.isEmpty()) {
      throw new BufferEmptyError();
    }

    const value = this.buffer[this.start];
    this.buffer[this.start] = null;
    this.start = (this.start + 1) % this.size;
    this.full = false;
    return value;
  }

  forceWrite(value) {
    if (this.full) {
      this.buffer[this.start] = value;
      this.start = (this.start + 1) % this.size;
      this.end = (this.end + 1) % this.size;
    } else {
      this.write(value);
    }
  }

  clear() {
    this.buffer.fill(null);
    this.start = 0;
    this.end = 0;
    this.full = false;
  }

  isEmpty() {
    return !this.buffer.some((val) => val !== null);
  }

  isFull() {
    return this.full;
  }
}

export default CircularBuffer;

export class BufferFullError extends Error {
  constructor() {
    super("Buffer is full");
    this.name = "BufferFullError";
  }
}

export class BufferEmptyError extends Error {
  constructor() {
    super("Buffer is empty");
    this.name = "BufferEmptyError";
  }
}
